dll/cpl/sysdm/environment.c

```diff
@@ -45,97 +145,423 @@ GetSelectedListViewItem(HWND hwndListView)
-                        if (p && _tcschr(++p, _T('%')))
-                        {
-                            VarData->dwType = REG_EXPAND_SZ;
-                            VarData->lpCookedValue = GlobalAlloc(GPTR, 2 * MAX_PATH * sizeof(TCHAR));
+                    p = _tcschr(DlgData->VarData->lpRawValue, _T('%'));
+                    if (p && _tcschr(++p, _T('%')))
+                    {
+                        DlgData->VarData->dwType = REG_EXPAND_SZ;
+                        DlgData->VarData->lpCookedValue = GlobalAlloc(GPTR, 2 * MAX_PATH * sizeof(TCHAR));
 
-                            ExpandEnvironmentStrings(VarData->lpRawValue,
-                                                     VarData->lpCookedValue,
-                                                     2 * MAX_PATH);
-                        }
-                        else
-                        {
-                            VarData->dwType = REG_SZ;
-                            VarData->lpCookedValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
-                            _tcscpy(VarData->lpCookedValue, VarData->lpRawValue);
-                        }
+                        ExpandEnvironmentStrings(DlgData->VarData->lpRawValue,
+                                                 DlgData->VarData->lpCookedValue,
+                                                 2 * MAX_PATH);
+                    }
+                    else
+                    {
+                        DlgData->VarData->dwType = REG_SZ;
+                        DlgData->VarData->lpCookedValue = GlobalAlloc(GPTR, (dwValueLength + 1) * sizeof(TCHAR));
+                        _tcscpy(DlgData->VarData->lpCookedValue, DlgData->VarData->lpRawValue);
                     }
+
                     EndDialog(hwndDlg, 1);
                     return TRUE;
+                }
 
                 case IDCANCEL:
+                    if (IsWindow(DlgData->hEditBox))
+                    {
+                        ItemEditEnds(hwndDlg, DlgData);
+                    }
+
                     EndDialog(hwndDlg, 0);
                     return TRUE;
+
+                case IDC_BUTTON_BROWSE_FILE:
+                {
+                    if (IsWindow(DlgData->hEditBox))
+                    {
+                        ItemEditEnds(hwndDlg, DlgData);
+                    }
+
+                    BrowseRequiredFile(hwndDlg);
+                    break;
+                }
+
+                case IDC_BUTTON_BROWSE_FOLDER:
+                {
+                    if (IsWindow(DlgData->hEditBox))
+                    {
+                        ItemEditEnds(hwndDlg, DlgData);
+                    }
+
+                    BrowseRequiredFolder(hwndDlg, DlgData);
+                    break;
+                }
+
+                case IDC_BUTTON_DELETE:
+                {
+                    if (IsWindow(DlgData->hEditBox))
+                    {
+                        ItemEditEnds(hwndDlg, DlgData);
+                    }
```

Comment:
I would assume that the edit box has a lost focus or validate message, so that you don't have to litter your code in all other code paths with checks like this?